File: twister2/python-support/src/main/java/edu/iu/dsc/tws/python/tset/fn/PartitionFunctions.java
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
package edu.iu.dsc.tws.python.tset.fn;

import java.util.Set;

import edu.iu.dsc.tws.api.tset.fn.LoadBalancePartitioner;
import edu.iu.dsc.tws.api.tset.fn.PartitionFunc;
import edu.iu.dsc.tws.python.processors.PythonClassProcessor;

public final class PartitionFunctions extends TFunc<PartitionFunc> {

  private static final PartitionFunctions INSTANCE = new PartitionFunctions();

  private PartitionFunctions() {

  }

  static PartitionFunctions getInstance() {
    return INSTANCE;
  }

  public LoadBalancePartitioner loadBalanced() {
    return new LoadBalancePartitioner();
  }

  /**
   * This method builds a partition function
   *
   * @param pyBinary serialized python class
   */
  @Override
  public PartitionFunc build(byte[] pyBinary) {
    PythonClassProcessor partitionPython = new PythonClassProcessor(pyBinary);

    return new PartitionFunc() {
      @Override
      public void prepare(Set sources, Set destinations) {
        partitionPython.invoke("prepare", sources, destinations);
      }

      @Override
      public int partition(int sourceIndex, Object val) {
        return (Integer) partitionPython.invoke("partition", sourceIndex, val);
      }

      @Override
      public void commit(int source, int partition) {
        partitionPython.invoke("commit", source, partition);
      }
    };
  }
}

File: twister2/python-support/src/main/python/twister2/tset/fn/factory/PartitionFunctions.py
import cloudpickle as cp

from twister2.tset.fn.PartitionFunc import PartitionFunc


class PartitionFunctions:

    def __init__(self, java_ref, env):
        self.__java_ref = java_ref
        self.__env = env

    @property
    def load_balanced(self):
        return JavaWrapper(self.__java_ref.loadBalanced())

    def build(self, partition_func: PartitionFunc):
        # send python dump to java -> create a java object in JVM -> get the ref back
        return self.__java_ref.build(cp.dumps(partition_func))

    def to_java_ref(self, partition_func: PartitionFunc):
        if partition_func.pre_defined:
            return partition_func.java_ref()
        else:
            return self.build(partition_func)


class JavaWrapper(PartitionFunc):

    def prepare(self, sources: set, destinations: set):
        self.__java_ref.prepare(sources, destinations)

    def partition(self, source_index: int, val) -> int:
        return self.__java_ref.partition(source_index, val)

    def commit(self, source_index: int, partition: int) -> int:
        return self.__java_ref.commit(source_index, partition)

    def pre_defined(self) -> bool:
        return True

    def java_ref(self):
        return self.__java_ref

    def __init__(self, java_ref):
        super().__init__()
        self.__java_ref = java_ref
